prover/src/lib.rs

```diff
@@ -201,32 +201,49 @@ pub trait Prover {
     // PROVIDED METHODS
     // --------------------------------------------------------------------------------------------
 
-    /// Builds the GKR proof. If the [`Air`] doesn't use a GKR proof, leave unimplemented.
+    /// Builds and returns the auxiliary trace.
     #[allow(unused_variables)]
     #[maybe_async]
-    fn generate_gkr_proof<E>(
+    fn build_aux_trace<E>(
         &self,
         main_trace: &Self::Trace,
-        public_coin: &mut Self::RandomCoin,
-    ) -> (ProverGkrProof<Self>, GkrRandElements<E>)
+        aux_rand_elements: &AuxRandElements<E>,
+    ) -> ColMatrix<E>
     where
         E: FieldElement<BaseField = Self::BaseField>,
     {
-        unimplemented!("`Prover::generate_gkr_proof` needs to be implemented when the auxiliary trace has a Lagrange kernel column.")
+        unimplemented!("`Prover::build_aux_trace` needs to be implemented when the trace has an auxiliary segment.")
     }
 
     /// Builds and returns the auxiliary trace.
     #[allow(unused_variables)]
     #[maybe_async]
-    fn build_aux_trace<E>(
+    fn build_aux_trace_wrapper<E>(
         &self,
+        air: &Self::Air,
         main_trace: &Self::Trace,
         aux_rand_elements: &AuxRandElements<E>,
     ) -> ColMatrix<E>
     where
         E: FieldElement<BaseField = Self::BaseField>,
     {
-        unimplemented!("`Prover::build_aux_trace` needs to be implemented when the trace has an auxiliary segment.")
+        let mut aux_trace = self.build_aux_trace(main_trace, aux_rand_elements);
+
+        if let Some(lagrange_randomness) = aux_rand_elements.lagrange() {
+            let evaluator = air.get_logup_gkr_evaluator::<E::BaseField>();
+            let lagrange_col = build_lagrange_column(lagrange_randomness);
+            let s_col = build_s_column(
+                main_trace,
+                aux_rand_elements.gkr_data().expect("should not be empty"),
+                &evaluator,
+                &lagrange_col,
+            );
+
+            aux_trace.merge_column(s_col);
+            aux_trace.merge_column(lagrange_col);
+        }
+
+        aux_trace
     }
```

Comment:
Why do we need to have this wrapper? Could the following work?

- We have `build_aux_trace()` function (as we do now) and this function gets called when appropriate directly from  the `generate_proof()` function.
- We have a separate function called something like `build_gkr_trace_columns()` and this function either returns two columns or appends the columns to a passed in auxiliary trace.
  - I also wonder if maybe this function can be somehow combined with `prove_gkr()` function.

Reply:
Implemented your suggestion